In [115]:
import numpy as np
import pandas as pd

### step 1
gather and understandget insights about the entire dataset and columns etc , idntify null values type of columns (cataegorical , numerical ,  to keep , drop )

In [116]:
df = pd.read_csv("Datasets/titanic.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


from info() we can understand that there were total 11 columns and 891 entries so basicallly (891,11)

# EDA & Data understanding

In [117]:
df.sample(10)

PassengerId  Survived  Pclass  \
513          514         1       1   
217          218         0       2   
287          288         0       3   
83            84         0       1   
324          325         0       3   
92            93         0       1   
320          321         0       3   
867          868         0       1   
246          247         0       3   
889          890         1       1   

                                               Name     Sex   Age  SibSp  \
513  Rothschild, Mrs. Martin (Elizabeth L. Barrett)  female  54.0      1   
217                    Jacobsohn, Mr. Sidney Samuel    male  42.0      1   
287                            Naidenoff, Mr. Penko    male  22.0      0   
83                          Carrau, Mr. Francisco M    male  28.0      0   
324                        Sage, Mr. George John Jr    male   NaN      8   
92                      Chaffee, Mr. Herbert Fuller    male  46.0      1   
320                              Dennis, Mr. Samuel    male  22.0      0   
867            Roebling, Mr. Washington Augustus II    male  31.0      0   
246           Lindahl, Miss. Agda Thorilda Viktoria  female  25.0      0   
889                           Behr, Mr. Karl Howell    male  26.0      0   

     Parch       Ticket     Fare Cabin Embarked  
513      0     PC 17603  59.4000   NaN        C  
217      0       243847  27.0000   NaN        S  
287      0       349206   7.8958   NaN        S  
83       0       113059  47.1000   NaN        S  
324      2     CA. 2343  69.5500   NaN        S  
92       0  W.E.P. 5734  61.1750   E31        S  
320      0    A/5 21172   7.2500   NaN        S  
867      0     PC 17590  50.4958   A24        S  
246      0       347071   7.7750   NaN        S  
889      0       111369  30.0000  C148        C

### Findings till now
 - we dont need columns that has nothing to do with predictions like name , passengerId , ticket & cabin . drop them
 - also there is need of imputation in age , embarked column with mean and most freq as worthy candidates resp. (this is preprocessing task though)

In [118]:
df = df.drop(columns = ["PassengerId" , "Name" , "Ticket" , "Cabin"])
df

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.0      1      0   7.2500        S
1           1       1  female  38.0      1      0  71.2833        C
2           1       3  female  26.0      0      0   7.9250        S
3           1       1  female  35.0      1      0  53.1000        S
4           0       3    male  35.0      0      0   8.0500        S
..        ...     ...     ...   ...    ...    ...      ...      ...
886         0       2    male  27.0      0      0  13.0000        S
887         1       1  female  19.0      0      0  30.0000        S
888         0       3  female   NaN      1      2  23.4500        S
889         1       1    male  26.0      0      0  30.0000        C
890         0       3    male  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

In [119]:
# now lets identify if the data was balanced or skewed and obvoisly we need to check it for Survived column
df.value_counts("Survived", normalize = True)

Survived
0    0.616162
1    0.383838
Name: proportion, dtype: float64

it is still balanced so we will use accuracy_score metrics

## Classification Metrics – One-Line Interpretations

**Accuracy**
- Measures overall correctness of the model; suitable when classes are reasonably balanced and all errors have similar cost.

**Precision**
- Measures how trustworthy positive predictions are; used when false positives are costly and wrong alarms must be minimized.

**Recall**
- Measures how well the model captures all actual positive cases; used when false negatives are costly and missing positives is unacceptable.

**F1-Score**
- Harmonic mean of precision and recall; used when both false positives and false negatives matter and class imbalance exists.

In [120]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(df.drop(columns = ["Survived"]) , df["Survived"] , test_size = 0.3 )

In [121]:
df.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [122]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(2)
memory usage: 55.8+ KB


# transformer
what we will do is that we will create differnt transformer for various steps like 
- imputing missing values for age and embarked
- ohe for sex and embarked
- scaling so that every column gets to a specific range
- feature selection
- implement DT

In [123]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.impute import SimpleImputer

In [124]:
#for imputing missing values
trf1 = ColumnTransformer([
        ("impute_age" , SimpleImputer(strategy = "mean") , [2] ) ,
        ("impute_embarked" , SimpleImputer(strategy = "most_frequent" ) , [6])
    ],
    remainder = "passthrough"
)

In [125]:
#for one hot encoding of sex and embarked
trf2 = ColumnTransformer(
    [
        ("ohe_embarked_sex" , OneHotEncoder(sparse_output = False , handle_unknown = "ignore") , [2 , 6])
    ]
)

In [126]:
#for scaling 
trf3 = ColumnTransformer(
    [
        ("scaling" , MinMaxScaler() ,[0, 2, 3, 4, 5])
    ]
)

In [127]:
#for feature selection however not required here but for learning purpose only
from sklearn.feature_selection import SelectKBest , chi2
trf4 = SelectKBest(score_func = chi2 , k = 5 )

In [128]:
# for model implementations
from sklearn.tree import DecisionTreeClassifier
trf5 = DecisionTreeClassifier()

# pipeline
we will create a pipelines that connects this chains or transformers

In [129]:
from sklearn.pipeline import Pipeline , make_pipeline
pipe = Pipeline ([
    ("trf1" , trf1),
    ("trf2" , trf2),
    ("trf3" , trf3),
    ("trf4" , trf4),
    ("trf5" , trf5) ]
)

Pipeline vs make_pipeline (scikit-learn)

Pipeline:
- You manually assign step names
- More control over naming

Syntax:
Pipeline([
    ("scaler", StandardScaler()),
    ("model", LogisticRegression())
])

make_pipeline:
- Step names are auto-generated
- Less verbose, quicker to write

Syntax:
make_pipeline(
    StandardScaler(),
    LogisticRegression()
)


# why pipe.fit() and not pipe.fit_transform()
so basically if you are using a training a model as transformer step use fit so you can predict 

notes for reference 
- fit() is required to train the final estimator (model)
- Models do not have a transform() method
- fit_transform() works only when the last step is a transformer
- fit() enables predict(), score(), etc.

Rule:
If the pipeline ends with a model → use fit()
If the pipeline ends with a transformer → use fit_transform()


In [130]:
pipe.fit(X_train,y_train)

Pipeline(steps=[('trf1',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('impute_age', SimpleImputer(),
                                                  [2]),
                                                 ('impute_embarked',
                                                  SimpleImputer(strategy='most_frequent'),
                                                  [6])])),
                ('trf2',
                 ColumnTransformer(transformers=[('ohe_embarked_sex',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False),
                                                  [2, 6])])),
                ('trf3',
                 ColumnTransformer(transformers=[('scaling', MinMaxScaler(),
                                                  [0, 2, 3, 4, 5])])),
                ('trf4',
                 SelectKBest(k=5,
                             score_func=<function chi2 at 0x000001CE85488EA0>)),
                ('trf5', DecisionTreeClassifier())])

In [131]:
pipe.named_steps

{'trf1': ColumnTransformer(remainder='passthrough',
                   transformers=[('impute_age', SimpleImputer(), [2]),
                                 ('impute_embarked',
                                  SimpleImputer(strategy='most_frequent'),
                                  [6])]),
 'trf2': ColumnTransformer(transformers=[('ohe_embarked_sex',
                                  OneHotEncoder(handle_unknown='ignore',
                                                sparse_output=False),
                                  [2, 6])]),
 'trf3': ColumnTransformer(transformers=[('scaling', MinMaxScaler(), [0, 2, 3, 4, 5])]),
 'trf4': SelectKBest(k=5, score_func=<function chi2 at 0x000001CE85488EA0>),
 'trf5': DecisionTreeClassifier()}

In [132]:
pipe.named_steps["trf1"].transformers

[('impute_age', SimpleImputer(), [2]),
 ('impute_embarked', SimpleImputer(strategy='most_frequent'), [6])]

In [135]:
pipe.named_steps["trf1"].transformers[0]

('impute_age', SimpleImputer(), [2])

In [138]:
pipe.named_steps["trf1"].transformers[0][1].indicator_

AttributeError: 'SimpleImputer' object has no attribute 'indicator_'

# Machine Learning Workflow / Lifecycle (Reference Guide)

> **Purpose:**  
> This document acts as a reusable mental + practical framework for approaching any machine learning problem.
> The goal is not just to train a model, but to make *correct, explainable, and justified decisions* at every step.

---

## 1. Problem Framing

### 1.1 Define the Task
- What kind of ML problem is this?
  - Classification / Regression / Clustering / etc.
- If classification:
  - Binary or multiclass?
- What exactly should the model predict?

> Always state the task in **one clear sentence**.

---

### 1.2 Decision Context
- How will the model output be used?
  - Human decision support?
  - Fully automated decision?
- Is interpretability important?
- Is this a low-risk or high-risk decision?

These answers influence:
- Model choice
- Evaluation metric
- Complexity tolerance

---

## 2. Data Understanding

### 2.1 Row-Level Meaning
- What does **one row** represent?
  - One person?
  - One flower?
  - One transaction?

If this is unclear, stop and clarify before proceeding.

---

### 2.2 Column-Level Meaning
For each feature:
- What does it represent?
- Numeric or categorical?
- Measured or derived?
- Continuous or discrete?

⚠️ Never treat a column as “just a number” without understanding its meaning.

---

## 3. Exploratory Data Analysis (EDA)

EDA is about **validating assumptions**, not making plots for aesthetics.

---

### 3.1 Structural Checks
- Dataset shape
- Data types
- Missing values
- Duplicate rows

Key question:
> Is this dataset clean by design, or messy by nature?

---

### 3.2 Target Variable Analysis (Do This Early)
- Class distribution (for classification)
- Imbalance issues
- Rare classes

This step directly affects:
- Metric selection
- Sampling strategy
- Baseline expectations

---

### 3.3 Feature Analysis
For each feature:
- Distribution shape (normal, skewed, bounded)
- Scale range
- Presence of outliers

For feature–target relationship:
- Are classes separable?
- Linear or nonlinear patterns?
- Strong overlap or clear boundaries?

This is where **model intuition starts forming**.

---

## 4. Data Preprocessing (Decision-Driven)

Every preprocessing step must answer **WHY**, not just HOW.

---

### 4.1 Missing Values
Questions to ask:
- Why are values missing?
- Random or systematic?
- Should missingness itself be a signal?

Imputation strategy depends on:
- Feature meaning
- Distribution
- Model sensitivity

---

### 4.2 Scaling
Ask before scaling:
- Does the model rely on distances or gradients?
- Are feature scales very different?

General intuition:
- Distance / gradient-based models → scaling matters
- Tree-based models → scaling usually does not matter

---

### 4.3 Encoding
- Nominal vs ordinal categories
- Does numeric order have semantic meaning?

Avoid blind one-hot encoding.

---

## 5. Train–Test Strategy

Decide before modeling:
- Train–test split or cross-validation?
- Stratified split needed?
- Any data leakage risk?

Splitting strategy must reflect:
- Dataset size
- Target distribution
- Real-world usage scenario

---

## 6. Baseline Model (Mandatory Step)

Purpose of a baseline:
- Establish a performance floor
- Detect leakage early
- Build intuition

Examples:
- Logistic Regression
- k-NN
- Naive Bayes
- Simple Decision Tree

If a complex model barely beats the baseline → investigate the data, not the model.

---

## 7. Model Selection

Choose models based on:
- Dataset size
- Feature type
- Linearity vs non-linearity
- Interpretability requirements

You should be able to justify:
> “I chose this model because …”

---

## 8. Evaluation Metrics

Accuracy is a **choice**, not a default.

Ask:
- Cost of false positives vs false negatives?
- Class imbalance?
- Probabilities vs hard labels needed?

Metric must align with **problem cost**, not convenience.

---

## 9. Error Analysis

After evaluation:
- Which samples are misclassified?
- Any pattern in errors?
- Certain classes confused more than others?

This step guides:
- Feature engineering
- Model improvement
- Next experiment direction

---

## 10. Iteration & Insights

Always conclude with:
- What worked?
- What failed?
- What would I try next?

Even small datasets should produce insights.

---

> **Reminder:**  
> Machine Learning is not about finding the “best model”.
> It is about making **correct, explainable, and justified decisions** at every step.
